utilities/checkpoint/checkpoint_impl.cc

```diff
@@ -101,15 +107,54 @@ Status CheckpointImpl::CreateCheckpoint(const std::string& checkpoint_dir,
     return Status::InvalidArgument("invalid checkpoint directory name");
   }
 
-  std::string full_private_path =
-      checkpoint_dir.substr(0, final_nonslash_idx + 1) + ".tmp";
-  ROCKS_LOG_INFO(
-      db_options.info_log,
-      "Snapshot process -- using temporary directory %s",
-      full_private_path.c_str());
+  std::string parsed_checkpoint_dir =
+      checkpoint_dir.substr(0, final_nonslash_idx + 1);
+  std::string full_private_path = parsed_checkpoint_dir + ".tmp";
+  ROCKS_LOG_INFO(db_options.info_log,
+                 "Snapshot process -- using temporary directory %s",
+                 full_private_path.c_str());
   CleanStagingDirectory(full_private_path, db_options.info_log.get());
   // create snapshot directory
   s = db_->GetEnv()->CreateDir(full_private_path);
+
+  // Remove the last `/`s if needed
+  std::string parsed_log_dir =
+      db_log_dir.empty()
+          ? ""
+          : db_log_dir.substr(0, db_log_dir.find_last_not_of('/') + 1);
+  std::string parsed_wal_dir =
+      wal_dir.empty() ? ""
+                      : wal_dir.substr(0, wal_dir.find_last_not_of('/') + 1);
+
+  // Info log files are not copied or linked, just update the option value.
+  std::string value_log_dir = parsed_log_dir == db_->GetName() ||
+                                      parsed_log_dir == parsed_checkpoint_dir
+                                  ? ""
+                                  : parsed_log_dir;
+
+  // If the wal_dir is empty, or the same as the source db dir, update the
+  // option value to the checkpoint dir.
+  std::string value_wal_dir;  // Option value to override
+  std::string new_wal_dir;    // The target location to copy/link WAL files
+  if (parsed_wal_dir.empty() || parsed_wal_dir == db_->GetName() ||
+      parsed_wal_dir == parsed_checkpoint_dir) {
+    value_wal_dir = parsed_checkpoint_dir;
+    new_wal_dir = full_private_path;  // Copy to the temp dir
+  } else {
+    value_wal_dir = parsed_wal_dir;
+    std::string prefix = parsed_checkpoint_dir + "/";
+    // If checkpoint_dir is parent of wal_dir, create the wal dir inside the tmp
+    // dir; otherwise, create it directly.
+    new_wal_dir =
+        parsed_wal_dir.rfind(prefix, 0) == 0
+            ? full_private_path + "/" + parsed_wal_dir.substr(prefix.size())
+            : parsed_wal_dir;
```

Comment:
Suppose the `checkpoint_dir = "/a/b/c"`, then `full_private_path  =  "/a/b/c.tmp"`. If `wal_dir = "/a/b/c/wal"`, (or in multiple levels of sub-dir in checkpoint_dir), we cannot copy/link WAL files directly to the specified /a/b/c/wal.

By design, the other code in this file checks /a/b/c does not exist, and creates /a/b/c.tmp. If it creates /a/b/c/wal for WAL files, the later renaming /a/b/c.tmp to /a/b/c will fail. 

So, if `wal_dir` has a prefix `/a/b/c/` (with the last /), it means the `checkpoint_dir` is an ancestor dir of `wal_dir`, then WAL files should be copied/linked into the temporary dir `full_private_path `.

Reply:
Since checkpointing does not copy or link info logs, we don't need to handle the above case for `db_log_dir`. But the same problem may exist if a parent dir of `db_log_dir` does not exist, the program will crash since it's not able to create dirs recursively.